### PR TITLE
CA-401 Cookie Policy

### PR DESCRIPTION
**Pull request by ljwboschker**
Show cookie message banner on every page, until the 'Accept' button has been clicked. Acceptance state is stored in a cookie on the domain the site is running on.

### CODE REVIEW

**Comment by logophobia** [comment]
It might be a good idea to use the gdpr middleware to block cookies until we get consent:

https://docs.microsoft.com/en-us/aspnet/core/security/gdpr?view=aspnetcore-2.1

If you upgrade to asp 2.1, then you also solved CA-416.

Also, we should pair this with google tag manager modifications, to block google analytics until we get consent:

https://www.metric.ai/blog/gdpr-cookies

And we should probably block:
* Social media
* Share / Like
* Youtube

Until we get consent.

**Comment by ljwboschker** [comment]
@logophobia can you take a look at the latest changes?

Following is _not_ loaded until the cookie policy is accepted:
* Google Tag Manager
* Share buttons (homepage, 2x on project page and 1x on thank you page)
* Social media login options

Did I miss something?

**Comment by logophobia** [comment]
Also, if you set all cookies to needs consent, then I don't think they'll be able to login at all. So maybe block all of the login?

File: CollAction/Views/Shared/_Layout.cshtml [hunk 1]
```diff
@@ -1,11 +1,20 @@
-﻿<!DOCTYPE html>
+﻿@using Microsoft.AspNetCore.Http.Features
+
+@{
+    var consentFeature = Context.Features.Get<ITrackingConsentFeature>();
+    var canTrack = consentFeature?.CanTrack ?? false;
```
Comment (logophobia) [inline CollAction/Views/Shared/_Layout.cshtml]: I don't think you need the null coalesce here, if the consent feature isn't set, it should be an error right?

**Comment by logophobia** [comment]
Maybe in version 2 we could have options, like don't allow anonymous tracking, but do allow stuff for the essentials (like login).

File: CollAction/Frontend/app/global/CookieMessage.tsx [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    };
+
+    renderMessage() {
+        return (
+            <div>
+                <div id="cookie-message-banner">
+                    <div className="container">
+                        <div className="row">
+                            <div className="col-xs-12">
+                                <p>Thank you for visiting our website! Please note that our website uses cookies to analyze and improve the performance of our website and to make social media integration possible. For more information on the use of cookies and privacy related matters, please see our <a href="home/privacy">Privacy and Cookies Policy</a>. By clicking "I accept" or by using this site, you consent to the use of cookies.</p>
```
Comment (logophobia) [inline CollAction/Frontend/app/global/CookieMessage.tsx]: I think this is still missing a few things. I thought this was a good summary: https://www.cookiebot.com/en/gdpr-cookies/

* User should be able to reject the non-essential cookies (for example: tracking, youtube)
* I think this should be a bit clearer, maybe a bullet point list with a link to the full privacy policy?

File: CollAction/Views/Account/Login.cshtml [hunk 3]
```diff
@@ -10,6 +11,8 @@
 
 @{
     ViewData["Title"] = Localizer["Log in"];
+    var consentFeature = Context.Features.Get<ITrackingConsentFeature>();
```
Comment (logophobia) [inline CollAction/Views/Account/Login.cshtml]: Feels like it should be part of the viewmodel.

Comment (ljwboschker) [reply CollAction/Views/Account/Login.cshtml]: I'm not sure.. I got this from the Microsoft documentation itself. 

Also, there are 3 parts (index, _layout and _cookieConsentPartial) where there is no view model. I could introduce one for the index page, but not for _layout and _cookieConsentPartial. So we'll end up with two ways where this is set.

File: CollAction/Frontend/app/global/CookieMessage.tsx (hunk 2 above)
Comment (ljwboschker) [reply CollAction/Frontend/app/global/CookieMessage.tsx]: @ronvdakker can you comment on this? See if we need different copy here.

Do we want to give users an option to either opt-in or opt-out of tracking cookies. Perhaps you can create a new story for that?

Comment (ronvdakker) [reply CollAction/Frontend/app/global/CookieMessage.tsx]: Yes, we can create a separate story for the opt in and opt out of essentials. 

Lets go ahead with this one for now though, and improve upon it later.

**Comment by ljwboschker** [comment]
I checked the site without consent (including login and admin pages) and it seems to still work. 

Non-essential cookies are TempData (only used on social login) and session state (which we don't use). So it should work. Nevertheless, some extra testing of the site with the banner visible is in order.

File: CollAction/Views/Account/Login.cshtml (hunk 3 above)
Comment (logophobia) [reply CollAction/Views/Account/Login.cshtml]: Alright, didn't see that. It's fine as it is then.

**Pull request merged**